File: frontend/src/lib/permissionKeys.js
/**
 * Permission key constants — MUST stay in sync with dotnet-Backend/PermissionKeys.cs.
 *
 * can(permissions, key) returns true if user has the key OR system.fullaccess.
 * CAUSE: Sidebar and buttons hide modules user cannot call anyway (403 from API if bypassed).
 */
export const PK = {
  systemFullAccess: "system.fullaccess",
  dashboardHome: "dashboard.home",
  dashboardPosts: "dashboard.posts",
  postsView: "posts.view",
  postsCreate: "posts.create",
  postsEditOwn: "posts.edit_own",
  postsDeleteOwn: "posts.delete_own",
  postsManageAll: "posts.manage_all",
  profilePassword: "profile.password",
  formSubmissionsOwn: "form.submissions_own",
  adminRoles: "admin.roles",
  adminPermissions: "admin.permissions",
  adminUsers: "admin.users",
  adminSubmissions: "admin.submissions",
};

export function can(permissions, key) {
  if (!Array.isArray(permissions) || permissions.length === 0) return false;
  if (permissions.includes(PK.systemFullAccess)) return true;
  return permissions.includes(key);
}

File: dotnet-Backend/PermissionKeys.cs
/*
 * PermissionKeys — canonical permission strings (also in frontend lib/permissionKeys.js).
 * CAUSE: Policies use "Perm:" + key; handler checks JWT "permission" claims OR system.fullaccess.
 */
public static class PermissionKeys
{
    public const string SystemFullAccess = "system.fullaccess";

    public const string DashboardHome = "dashboard.home";
    public const string DashboardPosts = "dashboard.posts";

    public const string PostsView = "posts.view";
    public const string PostsCreate = "posts.create";
    public const string PostsEditOwn = "posts.edit_own";
    public const string PostsDeleteOwn = "posts.delete_own";
    public const string PostsManageAll = "posts.manage_all";

    public const string ProfilePassword = "profile.password";
    public const string FormSubmissionsOwn = "form.submissions_own";

    public const string AdminRoles = "admin.roles";
    public const string AdminPermissions = "admin.permissions";
    public const string AdminUsers = "admin.users";
    public const string AdminSubmissions = "admin.submissions";
}
